### PR TITLE
Created `GRPOTrainerWithEval` subclass for different evaluation reward functions

### PR DESCRIPTION
**Pull request by jamesbraza**
This PR creates a `GRPOTrainer` subclass `GRPOTrainerWithEval` that adds support for optional `eval_reward_processing_classes`.

It should be backwards compatible with `GRPOTrainer`.

The only caveat here is I didn't comprehensively think about `args.reward_weights`.

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1173,3 +1184,115 @@ def create_model_card(
...
+            reward_processing_classes = self.train_reward_processing_classes
+
+        rewards_per_func = torch.zeros(len(prompts), len(reward_funcs), device=device)
+        for i, (reward_func, reward_processing_class) in enumerate(
+            zip(reward_funcs, reward_processing_classes, strict=True)
+        ):
+            if isinstance(reward_func, nn.Module):  # Module instead of PretrainedModel for compat with compiled models
+                reward_func_name = f"reward {reward_func.config._name_or_path.split('/')[-1]}"
+            else:
+                reward_func_name = reward_func.__name__
```
[the remark this answers is not in the document]
Comment (jamesbraza) [reply trl/trainer/grpo_trainer.py:1263]: Thank you, I made https://github.com/huggingface/trl/issues/3049 for this

**Comment by shirinyamani** [comment]
Have you tested this against the new multi-task reward_func setup? 
Cuz the `restrict=True` will not work with the current setup which allows reward_funcs to return `None`

**Comment by jamesbraza** [comment]
> Have you tested this against the new multi-task reward_func setup? Cuz the `restrict=True` will not work with the current setup which allows reward_funcs to return `None`

Hi @shirinyamani thanks for the comment, no we stopped rebasing atop `trl`'s `main` because we were getting broken too often.

Perhaps if we rebased for newer features in `trl`, splitting out the eval reward function (what this PR does) would have to change due to this `restrict=True` thing.
